app/grandchallenge/algorithms/views.py

```diff
@@ -335,6 +338,67 @@ def get_context_data(self, *args, **kwargs):
         return context
 
 
+NOT_IN_REGISTRY_ERROR = "NOT_IN_REGISTRY"
+
+
+class AlgorithmImageMarkDesired(
+    LoginRequiredMixin,
+    ObjectPermissionRequiredMixin,
+    SuccessMessageMixin,
+    FormView,
+):
+    permission_required = "algorithms.change_algorithm"
+    raise_exception = True
+    form_class = ImageMarkDesiredForm
+    success_message = "Image successfully updated."
+    template_name = "algorithms/mark_desired_version.html"
+
+    @cached_property
+    def algorithm(self):
+        return self.algorithm_image.algorithm
+
+    @cached_property
+    def algorithm_image(self):
+        return get_object_or_404(AlgorithmImage, pk=self.kwargs["pk"])
+
+    def get_permission_object(self):
+        return self.algorithm
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs.update({"image": self.algorithm_image})
+        return kwargs
+
+    def form_valid(self, form):
+        response = super().form_valid(form=form)
+        self.algorithm_image.mark_desired_version()
+        return response
+
+    def form_invalid(self, form):
+        super().form_invalid(form=form)
+        if form.has_error(NON_FIELD_ERRORS, code=NOT_IN_REGISTRY_ERROR):
+            self.algorithm_image.import_status = ImportStatusChoices.QUEUED
+            self.algorithm_image.save()
+            upload_to_registry_and_sagemaker.signature(
+                kwargs={
+                    "app_label": self.algorithm_image._meta.app_label,
+                    "model_name": self.algorithm_image._meta.model_name,
+                    "pk": self.algorithm_image.pk,
+                    "mark_as_desired": True,
+                }
+            ).apply_async()
+            messages.success(
+                self.request,
+                "Image validation and upload to registry in progress. It can take a while for this image to become active, please be patient.",
+            )
+        else:
+            messages.error(self.request, form.non_field_errors())
+        return HttpResponseRedirect(self.get_success_url())
```

Comment:
This is what I was referring to in my comment. This is still awkward, returning to the success url in form_invalid, but we want to always return to the image detail page and show success and error messages there.

Reply:
Why not show them on the page where the form is submitted?

Reply:
That is the ImageDetail page.

Reply:
We're using HTMX to send the post request (through a button on the ImageDetail page)

Reply:
I could try to use the image detail template as the template for the form view, a bit awkward but it might work.

Reply:
If HTMX is an issue here then don't use it, put the button on the algorithm image detail page instead. If it is an issue then we need to look at HTMX in more detail as this is fundamental to how REST works.